client/src/connection/itc/index.ts

```diff
@@ -0,0 +1,413 @@
+        );
+        this._shellProcess.kill();
+        this._shellProcess = undefined;
+
+        this._workDirectory = undefined;
+        this._runReject = undefined;
+        this._runResolve = undefined;
+      }
+      this.clearPassword();
+      resolve();
+    });
+  };
+
+  /**
+   * Cancels a running SAS program
+   */
+  public cancel = async () => {
+    this._shellProcess.stdin.write("$runner.Cancel()\n", async (error) => {
+      if (error) {
+        this._runReject(error);
+      }
+
+      await this.fetchLog();
+    });
+  };
+
+  /**
+   * Formats the SAS Options provided in the profile into a format
+   * that the shell process can understand.
+   * @param sasOptions SAS Options array from the connection profile.
+   * @returns a string  denoting powershell syntax for an array literal.
+   */
+  private formatSASOptions = (sasOptions: string[]): string => {
+    const optionsVariable = `@("${sasOptions.join(`","`)}")`;
+    return optionsVariable;
+  };
+
+  /**
+   * Flushes the SAS log in chunks of [chunkSize] length,
+   * writing each chunk to stdout.
+   */
+  private fetchLog = async (): Promise<void> => {
+    this._shellProcess.stdin.write(
+      `
+do {
+  $chunkSize = 32768
+  $log = $runner.FlushLog($chunkSize)
+  Write-Host $log
+} while ($log.Length -gt 0)\n
+  `,
+      this.onWriteComplete,
+    );
+  };
+
+  /**
+   * Handles stderr output from the powershell child process.
+   * @param chunk a buffer of stderr output from the child process.
+   */
+  private onShellStdErr = (chunk: Buffer): void => {
+    const msg = chunk.toString();
+    console.warn("shellProcess stderr: " + msg);
+    this.clearPassword();
+    this._runReject(
+      new Error(
+        "There was an error executing the SAS Program.\nSee console log for more details.",
+      ),
+    );
+    // If we encountered an error in setup, we need to go through everything again
+    if (/Setup error/.test(msg)) {
+      this._shellProcess.kill();
+      this._workDirectory = undefined;
+    }
+  };
+
+  /**
+   * Handles stdout output from the powershell child process.
+   * @param data a buffer of stdout output from the child process.
+   */
+  private onShellStdOut = (data: Buffer): void => {
+    const output = data.toString().trimEnd();
+    const outputLines = output.split(/\n|\r\n/);
+
+    outputLines.forEach((line: string) => {
+      if (!line) {
+        return;
+      }
+      if (!this._workDirectory && line.startsWith("WORKDIR=")) {
+        const parts = line.split("WORKDIR=");
+        this._workDirectory = parts[1].trim();
+        this._runResolve();
+        return;
+      }
+      if (!this.processLineCodes(line)) {
+        this._html5FileName =
+          line.match(/NOTE: .+ HTML5.* Body.+: (.+)\.htm/)?.[1] ??
+          this._html5FileName;
+        this._onLogFn?.([{ type: "normal", line }]);
+      }
+    });
+  };
+
+  private processLineCodes(line: string): boolean {
+    if (line.includes(LineCodes.RunEndCode)) {
```

Comment:
It looks like this condition will enter twice for each run. The log will contain the source
`%put --vscode-sas-extension-submit-end--;`
and then the output log
`--vscode-sas-extension-submit-end--`
So it will fetch result twice.

Reply:
Oh you're right. I changed this _back_ to endsWith in _some_ commit, but I'm guessing bouncing between Mac and Windows meant I loss this along the way. Updated in https://github.com/sassoftware/vscode-sas-extension/pull/592/commits/8522fbdadb666dbbbf1e191300776895ec15cc7e